#14851 techdocs

malin-mallan opened the pull request:
## Hey, I just made a Pull Request!

Update `.toMatchImageSnapshot` from `cypress-plugin-snapshots` to `.matchImage` from` cypress-plugin-visual-regression-diff` in techdocs-cli

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

realandersn reviewed (commented):
Hello, thanks for taking a stab at this!

While playing around with this I noticed that we have upgraded to cypress version 10 some while ago, and the configuration has changed quite a lot with this version. I think we could take the opportunity to try and adress those changes as well, if you'd be up for it?

I started looking here: https://github.com/FRSOURCE/cypress-plugin-visual-regression-diff#getting-started. While there I noticed that the config file has changed in cypress v 10 (from `cypress.json` to either `cypress.config.js` or `cypress.config.ts`, which led me to the cypress migration docs: https://docs.cypress.io/guides/references/migration-guide#Migrating-to-Cypress-version-10-0

I think you'll get a nice guide for what needs to be done if trying to run the tests also:
```bash
cd packages/techdocs-cli
yarn test:cypress
```

In any case I think you'd want to add the new package to `package.json` (`yarn add -D @frsource/cypress-plugin-visual-regression-diff`) and also remove the old `"cypress-plugin-snapshots": "^1.4.4"`.

malin-mallan commented:
> Hello, thanks for taking a stab at this!
> 
> While playing around with this I noticed that we have upgraded to cypress version 10 some while ago, and the configuration has changed quite a lot with this version. I think we could take the opportunity to try and adress those changes as well, if you'd be up for it?
> 
> I started looking here: https://github.com/FRSOURCE/cypress-plugin-visual-regression-diff#getting-started. While there I noticed that the config file has changed in cypress v 10 (from `cypress.json` to either `cypress.config.js` or `cypress.config.ts`, which led me to the cypress migration docs: https://docs.cypress.io/guides/references/migration-guide#Migrating-to-Cypress-version-10-0
> 
> I think you'll get a nice guide for what needs to be done if trying to run the tests also:
> 
> ```shell
> cd packages/techdocs-cli
> yarn test:cypress
> ```
> 
> In any case I think you'd want to add the new package to `package.json` (`yarn add -D @frsource/cypress-plugin-visual-regression-diff`) and also remove the old `"cypress-plugin-snapshots": "^1.4.4"`.

I can try to fix that as well!

malin-mallan commented:
I have tried to make the new changes and changed cypress.json to cypress.config.js and added what is said on https://github.com/FRSOURCE/cypress-plugin-visual-regression-diff#getting-started in cypress.config.js and index.

malin-mallan commented:
I have been looking on migrating to cypess 10 for a while now.
I have renamed `cypress.json` to `cypress.config.js` and made the changes in `cypress.config.js` and `index.js` according to https://github.com/FRSOURCE/cypress-plugin-visual-regression-diff#getting-started
I have two questions though that I’m not sure about: 
-	Should we use `initPlugin` in e2e or component?
-	What should be done with the things that was in `cypress.json` before? Should it be kept but rewrite to js or should it be removed? I don´t get any errors while running `yarn test:cypress` when they are removed but I got a message in the terminal saying “Missing baseUrl in compilerOptions. tsconfig-paths will be skipped”

realandersn commented:
> I have two questions though that I’m not sure about:
> 
> * Should we use `initPlugin` in e2e or component?
> * What should be done with the things that was in `cypress.json` before? Should it be kept but rewrite to js or should it be removed? I don´t get any errors while running `yarn test:cypress` when they are removed but I got a message in the terminal saying “Missing baseUrl in compilerOptions. tsconfig-paths will be skipped”

* I'm not entirely sure, but I think we can try `e2e`.
* Yeah, I think we want to keep the config for the tests to work. I guess it has to be migrated to the new v 10 structure.

realandersn commented on package.json:
```diff
@@ -57,6 +57,7 @@
     "@backstage/create-app": "workspace:*",
     "@backstage/repo-tools": "workspace:*",
     "@changesets/cli": "^2.14.0",
+    "@frsource/cypress-plugin-visual-regression-diff": "^3.2.4",
```
This needs to be added to the `package.json` in `packages/techdocs-cli` and not here:
```bash
cd packages/techdocs-cli
yarn add -D @frsource/cypress-plugin-visual-regression-diff
```
In that file, you'll also find the line `cypress-plugin-snapshots": "^1.4.4`, which needs to be removed.

malin-mallan commented:
@realandersn, would you like to take a look again?

jhaals commented on .changeset/cool-ladybugs-argue.md:
```diff
@@ -0,0 +1,8 @@
+'@backstage/core-components': minor
+'@backstage/create-app': minor
+'@techdocs/cli': minor
+'@backstage/plugin-dynatrace': minor
```
These packages does not need a changeset as there are no changes to those packages and cypress is just an internal test concern.

jhaals commented on .changeset/calm-dolphins-joke.md:
```diff
@@ -0,0 +1,5 @@
+'@techdocs/cli': minor
+---
+
+Change from cypress-plugin-snapshots to cypress-plugin-visual-regression-diff
```
Is cypress settings or configuration something that's shipped with this package? If not I don't think we need to include this changeset either.

jhaals reviewed (changes requested):
Thanks for looking at this @malin-mallan, really nice to see that massive yarn.lock cleanup if that's caused by these fixes! I put down a few comments, this PR also need to be updated against master to fix the conflicts :)

realandersn commented on packages/techdocs-cli/cypress.config.js:
```diff
@@ -0,0 +1,40 @@
+        resizeDevicePixelRatio: true,
+        threshold: 0.01,
+      },
+    },
```
This looks like config for the old package, I think we can just remove it. There's config for the new one as well (https://github.com/FRSOURCE/cypress-plugin-visual-regression-diff#configuration), but as far as I can tell the defaults should be fine.

realandersn commented on packages/techdocs-cli/cypress.config.js:
```diff
@@ -0,0 +1,40 @@
+      initPlugin(on, config);
+    },
+  },
+  excludeSpecPattern: ['**/__snapshots__/*', '**/__image_snapshots__/*'],
```
The `excludeSpecPattern` needs to be within the ´e2e´ block.

realandersn commented:
With the changes already proposed, the last step needed (I think 😅) is to change the structure according to: https://docs.cypress.io/guides/core-concepts/writing-and-organizing-tests#Folder-structure:
* Rename the `integrations` folder to `e2e` and the test files to `backstage_serve.cy.js` and `mkdocs_serve.cy.js`.
* Rename `support/index.js` to `support/e2e.js`.
* Remove the `plugins` folder and `index.js` inside it.

malin-mallan commented:
I think that all comments are fixed except for solving the merge conflict with yarn.lock. I get errors that I don't know how to solve when I try to solve the merge conflict. Can any of you help me with solving the conflict?

jhaals commented:
> I think that all comments are fixed except for solving the merge conflict with yarn.lock. I get errors that I don't know how to solve when I try to solve the merge conflict. Can any of you help me with solving the conflict?

If you haven't performed any `git merge` into this branch previously you should be able to rebase.
Not sure if you have an upstream remote already but if not you can `git remote add upstream https://github.com/backstage/backstage.git`

In case this causes more problems or simply does not work just forklift your changes to new branch that's created from a fresh master and make a new PR 😄 🤞

```
git fetch upstream
git rebase upstream/master
# once you hit a merge conflict run yarn to fix conflict in yarn.lock
yarn
git add yarn.lock
# continue until the next conflict.
git rebase --continue
# repeat until there are no more conflicts.
git push --force
```
